logging.c

```diff
@@ -180,13 +226,108 @@ static void owl_log_entry_free(void *data)
   }
 }
 
+#if GLIB_CHECK_VERSION(2, 32, 0)
+#else
+static void owl_log_entry_free_gfunc(gpointer data, gpointer user_data)
+{
+  owl_log_entry_free(data);
+}
+#endif
+
+static void owl_log_file_error(owl_log_entry *msg, int ret)
+{
+  owl_log_error("Unable to open file for logging: %s (file %s)",
+                g_strerror(ret),
+                msg->filename);
+}
+
+/* If we are deferring log messages, enqueue this entry for writing.
+ * Otherwise, try to write this log message, and, if it fails with
+ * EPERM, EACCES, or ETIMEDOUT, go into deferred logging mode and
+ * queue an admin message.  If it fails with anything else, display an
+ * error message, but do not go into deferred logging mode. */
+static void owl_log_eventually_write_entry(gpointer data)
+{
+  int ret;
+  owl_log_entry *msg = (owl_log_entry*)data;
+  if (defer_logs) {
+    owl_log_deferred_enqueue_message(msg->message, msg->filename);
+  } else {
+    ret = owl_log_try_write_entry(msg);
+    if (ret == EPERM || ret == EACCES || ret == ETIMEDOUT) {
+      defer_logs = true;
+      owl_log_error("Unable to open file for logging (%s): \n"
+                    "%s.  \n"
+                    "Consider renewing your tickets.  Logging has been \n"
+                    "suspended, and your messages will be saved.  To \n"
+                    "resume logging, use the command :flush-logs.\n\n",
+                    msg->filename,
+                    g_strerror(ret));
+      /* If we were not in deferred logging mode, either the queue should be
+       * empty, or we are attempting to log a message that we just popped off
+       * the head of the queue.  Either way, we should enqueue this message as
+       * the first message in the queue, rather than the last, so that we
+       * preserve message ordering. */
+      owl_log_deferred_enqueue_first_message(msg->message, msg->filename);
+    } else if (ret != 0) {
+      owl_log_file_error(msg, ret);
+    }
+  }
+}
+
+/* tries to write the deferred log entries */
+static void owl_log_write_deferred_entries(gpointer data)
+{
+  owl_log_entry *entry;
+  bool drop_failed_logs = *(bool *)data;
+  int ret;
+  int logged_message_count = 0;
+  bool all_succeeded = true;
+
+  defer_logs = false;
```

Comment:
Is this racey? It's written and read from two functions -- are they the same thread? Jason says in person that they are, but it'd be good to add a comment about that.